src/Oscoin/Protocol/Sync/RealWorld.hs

```diff
@@ -0,0 +1,169 @@
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE NumericUnderscores   #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-deprecations #-} -- Temporary
+{-# OPTIONS_GHC -fno-warn-redundant-constraints #-} -- Temporary
+module Oscoin.Protocol.Sync.RealWorld (
+      syncNode
+    , newSyncContextIO
+    , runSyncIO
+    ) where
+
+import           Oscoin.Prelude
+
+import qualified Oscoin.Consensus.Config as Consensus
+import           Oscoin.Crypto.Blockchain.Block (Block, BlockHash, Sealed)
+import           Oscoin.Crypto.Hash (HasHashing, Hash)
+import           Oscoin.Crypto.PubKey (PublicKey)
+import qualified Oscoin.Crypto.PubKey as Crypto
+import           Oscoin.P2P as P2P
+import           Oscoin.Protocol.Sync
+import           Oscoin.Storage.Block.Abstract (BlockStoreReader)
+import qualified Oscoin.Storage.Block.Abstract as BlockStore
+import           Oscoin.Telemetry.Trace
+import           Oscoin.Time (Duration, microseconds, seconds)
+import           Oscoin.Time.Chrono
+
+import           Codec.Serialise as CBOR
+import qualified Control.Concurrent.Async as Async
+import           Control.Retry
+                 (RetryPolicyM, RetryStatus, exponentialBackoff, retrying)
+import           Data.Hashable (Hashable)
+import qualified Data.HashSet as Set
+import qualified Network.Gossip.HyParView as Gossip
+import           Network.Gossip.IO.Peer (Peer(..))
+import qualified Network.HTTP.Client as HTTP
+
+{------------------------------------------------------------------------------
+  Scary IO implementation
+------------------------------------------------------------------------------}
+
+maxAllowedNetworkLatency :: Duration
+maxAllowedNetworkLatency = 1 * seconds
+
+newDataFetcherIO
+    :: forall c tx s.
+    ( Serialise (ProtocolResponse c tx s 'GetTip) )
+    => HTTP.Manager
+    -> DataFetcher c tx s IO
+newDataFetcherIO httpManager = DataFetcher $ \peer -> \case
+    -- This implementation is basically 'fetchTip' function from the spec.
+    SGetTip -> \() -> timed maxAllowedNetworkLatency $ do
+        let hn = addrHost . nodeHttpApiAddr $ peer
+        -- FIXME(adn) Support for https if needed.
+        rq0 <- HTTP.parseUrlThrow ("http://" <> toS (renderHost hn) <> "/blockchain/tip")
```

Comment:
We should make sure this only returns the header, since it can otherwise result in several megabites of download just to get the best height.

Reply:
At the moment we can't , as we don't have an equivalent `getBlockHeadersByHeight`, and I am OK for now to keep the dumb version up until we don't need to start supporting light clients.

See also https://github.com/oscoin/oscoin/issues/550#issuecomment-494331328 (TL;DR; Thomas and I accepted the compromise to delay this a little longer).

Reply:
(Deleted my previous comment as it wasn't very meaningful). At the moment we can't, as we don't have a `getTipHeader` HTTP API. I feel the case of "give me the header" vs "give me the full block" is common enough that perhaps we should be thinking having our HTTP API accept a parameter (for each request) where you can specify which "view" you want over a block (i.e. either `light` or `full`). This way at least we wouldn't duplicate all the endpoints.